### PR TITLE
{cmd,internal}/contour: add gRPC-Web support

### LINKED ISSUE #300: Support gRPC-Web
Proposals:
1. One annotation `contour.heptio.com/http-filter.{name}` and just support those filters that require no config parameters, i.e `grpc_web` :)
2. Two annotations `contour.heptio.com/http-filter.{name}` and another one for the configuration `contour.heptio.com/http-filter.{http_filter_name}-config`
3. One annotation to enable it like the above and then another that points to a config map with custom http filter parameters.

Do you agree that it's preferable to put this on a `Ingress`?

In case there are no objections I would like to go with the first one. Start small and extend this feature in further PRs.

/cc @davecheney 

Reference:
https://www.envoyproxy.io/docs/envoy/v1.6.0/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto.html#config-filter-network-http-connection-manager-v2-httpfilter

### PR DESCRIPTION
**Pull request by glerchundi**
Closes #300

Signed-off-by: Gorka Lerchundi Osa <glertxundi@gmail.com>

### CODE REVIEW

**Comment by glerchundi** [comment]
I've created a new cli parameter so that people can choose if gRPC-Web is enabled by default or not. Let me know if you prefer me to completely remove this cli arg and go through those 'sane defaults'.

PTAL @davecheney

**Comment by glerchundi** [comment]
Although I signed it off i think status reporting got stuck.

File: internal/contour/listener.go [hunk 1]
```diff
@@ -262,7 +269,17 @@ func tlscontext(secret *v1.Secret, alpnprotos ...string) *auth.DownstreamTlsCont
 	}
 }
 
-func httpfilter(routename string) listener.Filter {
+func httpfilter(routename string, useGrpcWeb bool) listener.Filter {
```
Comment (davecheney) [inline internal/contour/listener.go]: Please don't add a parameter here. Add a new helper, say `grpcwebfilter` which sets the appropriate settings.

File: internal/contour/listener.go [hunk 2]
```diff
@@ -54,11 +55,17 @@ type ListenerCache struct {
...
+	// UseProxyProto configures all listeners to expect a PROXY protocol
 	// V1 header on new connections.
 	// If not set, defaults to false.
 	UseProxyProto bool
 
+	// UseGrpcWeb enables a filter when bridging of a gRPC-Web client to a
+	// compliant gRPC server by following
+	// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md.
+	// If not set, defaults to false.
+	UseGrpcWeb bool
```
Comment (davecheney) [inline internal/contour/listener.go]: There needs to be a test of this value set to true.

**Comment by davecheney** [comment]
Thanks for working on this. I'm sorry for the delay, and there will be more as I will be on leave next week. 

Sadly I don't want to merge this as I do not think it is reasonable to dedicate all listeners to grpc-web. I'm fine with a per ingress annotation, but not a flag that alerters the behaviour of the listener.

I'm going to close this PR as per the Contributing documentation. Please let's keep talking about how to implement this. I'm ok with enabling this setting globally with a default of true if it applies transparently -- it's not clear from the reading of the envoy or grpc-web documentation if this is true  -- and there is no reasonable reason to not leave it turned on for all users. This is the same argument i'll make for enabling gzip everywhere without an option to turn it off.

**Comment by glerchundi** [comment]
> Sadly I don't want to merge this as I do not think it is reasonable to dedicate all listeners to grpc-web. I'm fine with a per ingress annotation, but not a flag that alerters the behaviour of the listener.

Sad news :(

> I'm ok with enabling this setting globally with a default of true if it applies transparently -- it's not clear from the reading of the envoy or grpc-web documentation if this is true -- and there is no reasonable reason to not leave it turned on for all users.

Ok, lets discuss about this in #300 then. Although I thought it was already discussed.

**Pull request closed without merging**